modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RatedDocumentKey.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.rankeval;
+
+import org.elasticsearch.action.support.ToXContentToBytes;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class RatedDocumentKey extends ToXContentToBytes implements Writeable {
+    public static final ParseField DOC_ID_FIELD = new ParseField("doc_id");
+    public static final ParseField TYPE_FIELD = new ParseField("type");
+    public static final ParseField INDEX_FIELD = new ParseField("index");
+
+    private static final ObjectParser<RatedDocumentKey, RankEvalContext> PARSER = new ObjectParser<>("ratings", RatedDocumentKey::new);
+
+    static {
+        PARSER.declareString(RatedDocumentKey::setIndex, INDEX_FIELD);
+        PARSER.declareString(RatedDocumentKey::setType, TYPE_FIELD);
+        PARSER.declareString(RatedDocumentKey::setDocId, DOC_ID_FIELD);
+    }
+    
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(INDEX_FIELD.getPreferredName(), index);
+        builder.field(TYPE_FIELD.getPreferredName(), type);
+        builder.field(DOC_ID_FIELD.getPreferredName(), docId);
+        builder.endObject();
+        return builder;
+    }
+    
+    // TODO instead of docId use path to id and id itself
+    private String docId;
+    private String type;
+    private String index;
+
+    public RatedDocumentKey() {}
```

Comment:
Same here, I think the ctor should contain the mandatory arguments and those should be immutable later. Again, this means for parsing we need ConstructingObjectParser.